src/hotspot/share/opto/reachability.cpp

```diff
@@ -0,0 +1,554 @@
+ * interferes with referent's live range.
+ *
+ * It is tempting to directly attach referents to interfering safepoints right from the beginning, but it
+ * doesn't play well with some optimizations C2 does (e.g., during loop-invariant code motion a safepoint
+ * can become interfering once a load is hoisted).
+ *
+ * Instead, reachability representation transitions through multiple phases:
+ *   (0) initial set of RFs is materialized during parsing (as a result of
+ *       Reference.reachabilityFence intrinsification);
+ *   (1) optimization pass during loop opts eliminates redundant RF nodes and
+ *       moves the ones with loop-invariant referents outside (after) loops;
+ *   (2) after loop opts are over, RF nodes are eliminated and their referents are transferred to
+ *       safepoint nodes (appended as edges after debug info);
+ *   (3) during final graph reshaping, referent edges are removed from safepoints and materialized as RF nodes
+ *       attached to their safepoint node (closely following it in CFG graph).
+ *
+ * Some implementation considerations.
+ *
+ * (a) It looks attractive to get rid of RF nodes early and transfer to safepoint-attached representation,
+ * but it is not correct until loop opts are done.
+ *
+ * Live ranges of values are routinely extended during loop opts. And it can break the invariant that
+ * all interfering safepoints contain the referent in their oop map. (If an interfering safepoint doesn't
+ * keep the referent alive, then it becomes possible for the referent to be prematurely GCed.)
+ *
+ * After loop opts are over, it becomes possible to reliably enumerate all interfering safe points and
+ * to ensure that the referent is present in their oop maps.
+ *
+ * (b) RF nodes may interfere with Register Allocator (RA). If a safepoint is pruned during macro expansion,
+ * it can make some RF nodes redundant, but we don't have information about their relations anymore to detect that.
+ * Redundant RF node unnecessarily extends referent's live range and increases register pressure.
+ *
+ * Hence, we eliminate RF nodes and transfer their referents to corresponding safepoints (phase #2).
+ * When safepoints are pruned, corresponding reachability edges also go away.
+ *
+ * (c) Unfortunately, it's not straightforward to stay with safepoint-attached representation till the very end,
+ * because information about derived oops is attached to safepoints in a similar way. So, for now RFs are
+ * rematerialized at safepoints before RA (phase #3).
+ */
+
+// RF is redundant for some referent oop when the referent has another user which keeps it alive across the RF.
+// In terms of dominance relation it can be formulated as "a referent has a user which is dominated by the redundant RF".
+// Until loop opts are over, only RF nodes are considered as usages (controlled by rf_only flag).
+static bool is_redundant_rf_helper(Node* ctrl, Node* referent, PhaseIdealLoop* phase, PhaseGVN& gvn, bool rf_only) {
```

Comment:
Nit: `_helper` is fine if it is used as some internal method, i.e. only `is_redundant_rf` uses `is_redundant_rf_helper`. But it seems you are using it from different places. Can you find a better name?